File: src/main/java/pl/androidcoder/solid/lsp/solve/collection/LimitedStorage.java
package pl.androidcoder.solid.lsp.solve.collection;

import pl.androidcoder.solid.lsp.collection.Storage;

import java.util.ArrayList;

public class LimitedStorage<T> implements Storage<T> {

    private int limit;

    private ArrayList<T> list = new ArrayList<>();

    public LimitedStorage(int itemLimit) {
        limit = itemLimit;
    }

    @Override
    public boolean addItem(T item) {
        if (list.size() >= limit)
            return false;
        else
            list.add(item);
        return true;
    }

    public void addNewItem(T item) throws LimitExceededException {
        if (!addItem(item)) throw new LimitExceededException();
    }
}

File: src/main/java/pl/androidcoder/solid/lsp/solve/collectionkotlin/LimitedStorage.kt
package pl.androidcoder.solid.lsp.solve.collectionkotlin

import pl.androidcoder.solid.lsp.collection.Storage
import pl.androidcoder.solid.lsp.collectionkotlin.LimitExceededException
import java.util.*

class LimitedStorage<T>(private val limit: Int) : Storage<T> {

    private val list = ArrayList<T>()

    override fun addItem(item: T): Boolean {
        if (list.size >= limit)
            return false
        else
            list.add(item)
        return true
    }

    fun addNewItem(item: T) {
        if (!addItem(item)) throw LimitExceededException()
    }
}